Improvements to timing accuracy

bts opened the pull request:
Hi Adam,

Here are a few changes I've found bring rate-limited actions closer to the target rate limit.

Before when I was using `PerExecution` and tried to do e.g. 500 HTTP requests per second, I would only get ~300/sec. With `PerInvocation` I would get around ~310/sec. After these changes, using either method, I get ~495/sec.

These changes:
- Use elapsed total/system clock time instead of elapsed CPU time. This adds a dependency on the `time` package.
- Use the time *before* the last run, not after, to calculate sleep time. This is important for `PerExecution` when `IO` actions take a nontrivial amount of time.
- Calculate by how much time `threadDelay` has overslept, and discount that extra time from future sleeps.

Let me know if you have any questions, or if there's anything I can change.

Thanks!
Brian

acw commented:
My one concern with this set of patches is the time 1.8 dependency. Based on [https://ghc.haskell.org/trac/ghc/wiki/Commentary/Libraries/VersionHistory this], it looks like >= 1.8 limits this library to GHC 8.2.1 or better. (Well, not totally, but you do set people up for some conflict headaches when we require new versions of base libraries.) I know that this library has been used as far back as GHC 7.10.3. Is there any way to weaken this requirement to allow for time 1.5.0.1 or better, instead?

bts commented:
Whoops sorry about that! I'll fix that now.

bts commented:
I switched from `Data.Time.Clock.System.getSystemTime` to `Data.Time.Clock.POSIX.getPOSIXTime`, which has been around since at least `time` `1.5.0.1`, and adjusted the range of the dependency to `>= 1.5.0.1 && < 1.9`.

acw commented:
Perfect, thanks!

acw commented:
I've bumped the version, and uploaded this to Hackage as 1.4.0. Thanks again!
